ext/vulnsrc/amzn/amzn.go

```diff
@@ -0,0 +1,343 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package amzn implements a vulnerability source updater using
+// ALAS (Amazon Linux Security Advisories).
+package amzn
+
+import (
+	"bufio"
+	"compress/gzip"
+	"encoding/xml"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/coreos/clair/database"
+	"github.com/coreos/clair/ext/versionfmt"
+	"github.com/coreos/clair/ext/versionfmt/rpm"
+	"github.com/coreos/clair/ext/vulnsrc"
+	"github.com/coreos/clair/pkg/commonerr"
+	"github.com/coreos/clair/pkg/httputil"
+)
+
+const (
+	amazonLinux1UpdaterFlag   = "amazonLinux1Updater"
+	amazonLinux1MirrorListURI = "http://repo.us-west-2.amazonaws.com/2018.03/updates/x86_64/mirror.list"
+	amazonLinux1Name          = "Amazon Linux 2018.03"
+	amazonLinux1Namespace     = "amzn:2018.03"
+	amazonLinux1LinkFormat    = "https://alas.aws.amazon.com/%s.html"
+
+	amazonLinux2UpdaterFlag   = "amazonLinux2Updater"
+	amazonLinux2MirrorListURI = "https://cdn.amazonlinux.com/2/core/latest/x86_64/mirror.list"
+	amazonLinux2Name          = "Amazon Linux 2"
+	amazonLinux2Namespace     = "amzn:2"
+	amazonLinux2LinkFormat    = "https://alas.aws.amazon.com/AL2/%s.html"
+	defaultUpdaterFlagValue   = ""
+)
+
+type updater struct {
+	UpdaterFlag   string
+	MirrorListURI string
+	Name          string
+	Namespace     string
+	LinkFormat    string
+}
+
+func init() {
+	// Register updater for Amazon Linux 2018.03.
+	amazonLinux1Updater := updater{
+		UpdaterFlag:   amazonLinux1UpdaterFlag,
+		MirrorListURI: amazonLinux1MirrorListURI,
+		Name:          amazonLinux1Name,
+		Namespace:     amazonLinux1Namespace,
+		LinkFormat:    amazonLinux1LinkFormat,
+	}
+	vulnsrc.RegisterUpdater("amzn1", &amazonLinux1Updater)
+
+	// Register updater for Amazon Linux 2.
+	amazonLinux2Updater := updater{
+		UpdaterFlag:   amazonLinux2UpdaterFlag,
+		MirrorListURI: amazonLinux2MirrorListURI,
+		Name:          amazonLinux2Name,
+		Namespace:     amazonLinux2Namespace,
+		LinkFormat:    amazonLinux2LinkFormat,
+	}
+	vulnsrc.RegisterUpdater("amzn2", &amazonLinux2Updater)
+}
+
+func (u *updater) Update(datastore database.Datastore) (vulnsrc.UpdateResponse, error) {
+	log.WithField("package", u.Name).Info("Start fetching vulnerabilities")
+
+	// Get the most recent updated date of the previous update.
+	flagValue, err := datastore.GetKeyValue(u.UpdaterFlag)
+	if err != nil {
+		return vulnsrc.UpdateResponse{}, err
+	}
+
+	if flagValue == "" {
```

Comment:
GetKeyValue does not return a `found` value, unlike FindKeyValueAndRollback. Instead, it returns the empty string when the key was not found. If the key is not found, I use a default flag value, which is also the empty string, coincidentally. I wrote this code to be explicit about this logic.

Reply:
a comment here explaining this is probably important, because i could totally see someone removing this code when they realize it's basically a noop